packages/frontend/component/src/ui/dnd/draggable.ts

```diff
@@ -0,0 +1,242 @@
+  position?: 'pointer-outside' | 'pointer-center' | 'native';
+}>;
+
+export const useDraggable = <D extends DNDData = DNDData>(
+  getOptions: () => DraggableOptions<D> = () => ({}),
+  deps: any[] = []
+) => {
+  const [dragging, setDragging] = useState<boolean>(false);
+  const [draggingOffset, setDraggingOffset] = useState<{
+    offsetX: number;
+    offsetY: number;
+    clientX: number;
+    clientY: number;
+    outWindow: boolean;
+  }>({ offsetX: 0, offsetY: 0, clientX: 0, clientY: 0, outWindow: false });
+  const [dropTarget, setDropTarget] = useState<
+    (DropTargetRecord & { data: D['dropTarget'] })[]
+  >([]);
+  const [customDragPreviewPortal, setCustomDragPreviewPortal] = useState<
+    React.FC<DraggableCustomDragPreviewProps>
+  >(() => () => null);
+
+  const dragRef = useRef<any>(null);
+  const dragHandleRef = useRef<any>(null);
+
+  const enableCustomDragPreview = useRef(false);
+  const enableDraggingPosition = useRef(false);
+  const enableDropTarget = useRef(false);
+  const enableDragging = useRef(false);
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const options = useMemo(getOptions, deps);
+
+  useEffect(() => {
+    if (!dragRef.current) {
+      return;
+    }
+
+    const windowEvent = {
+      dragleave: () => {
+        setDraggingOffset(state =>
+          state.outWindow === true ? state : { ...state, outWindow: true }
+        );
+      },
+      dragover: () => {
+        setDraggingOffset(state =>
+          state.outWindow === true ? { ...state, outWindow: false } : state
+        );
+      },
+    };
+
+    const cleanupDraggable = draggable({
+      element: dragRef.current,
+      dragHandle: dragHandleRef.current ?? undefined,
+      canDrag: draggableGet(options.canDrag),
+      getInitialData: draggableGet(options.data),
+      getInitialDataForExternal: draggableGet(options.dataForExternal),
+      onDragStart: () => {
+        if (enableDragging.current) {
+          setDragging(true);
+        }
+        if (enableDraggingPosition.current) {
+          document.body.addEventListener('dragleave', windowEvent.dragleave);
+          document.body.addEventListener('dragover', windowEvent.dragover);
+          setDraggingOffset({
+            offsetX: 0,
+            offsetY: 0,
+            clientX: 0,
+            clientY: 0,
+            outWindow: false,
+          });
+        }
+        if (enableDropTarget.current) {
+          setDropTarget([]);
+        }
+        if (dragRef.current) {
+          dragRef.current.dataset['dragging'] = 'true';
+        }
+      },
+      onDrop: () => {
+        if (enableDragging.current) {
+          setDragging(false);
+        }
+        if (enableDraggingPosition.current) {
+          window.removeEventListener('dragleave', windowEvent.dragleave);
+          window.removeEventListener('dragover', windowEvent.dragover);
+          setDraggingOffset({
+            offsetX: 0,
+            offsetY: 0,
+            clientX: 0,
```

Comment:
Initial client position shouldn't be `(0, 0)` ?

Reply:
I changed the initial clientXY in onDragStart to real mouse position